reggie/ingestion/preprocessor/west_virginia_preprocessor.py

```diff
@@ -1,87 +1,361 @@
+
+            # Create dict for encoding
+            array_encoding = {}
+            for k, row in df_elections.iterrows():
+                array_encoding[row["id_election"]] = {
+                    "index": k,
+                    "count": row["count"],
+                    # TODO: The documentation specifically says this should be MM/DD/YYYY,
+                    # but other files have the more standard YYYY-MM-DD
+                    # https://github.com/Voteshield/Inspector/wiki/Adding-a-State
+                    "date": row["dt_election"].strftime("%m/%d/%Y")
+                    # "name": row["Election_Name"]
+                }
+
+            # Create meta data
+            self.meta = {
+                "message": "{}_{}".format(config["state"], datetime.now().isoformat()),
+                "array_encoding": json.dumps(array_encoding),
+                "array_decoding": json.dumps(list(df_elections["id_election"])),
+            }
+
+        # Attach for testing and more direct access
+        # See: https://github.com/Voteshield/reggie/issues/50
+        self.processed_df = df_voters
+
+        # Create file from processed dataframe
         self.processed_file = FileItem(
             name="{}.processed".format(self.config["state"]),
-            io_obj=StringIO(df_voter.to_csv(encoding="utf-8", index=True)),
+            io_obj=StringIO(df_voters.to_csv(encoding="utf-8", index=True)),
             s3_bucket=self.s3_bucket,
         )
+
+    def clean_party_value(self, value):
+        """
+        Party is very variable, maybe allows for write-in,
+        so we try to clean up, but default to unknown.
+        """
+        value = str(value).strip().lower()
+        lookup = {
+            "democrat": re.compile("democrat"),
+            "republican": re.compile("republican"),
+            "libertarian": re.compile("libertarian"),
+            "independent": re.compile("independent"),
+            "green": re.compile("green"),
+            "unaffiliated": re.compile("(unaffiliated|no.*affiliation)"),
+        }
+
+        for k, v in lookup.items():
+            if v.match(value):
+                return k
+
+        return "unknown"
+
+    def clean_challenged_value(self, value):
+        true_regex = re.compile("^y$", re.IGNORECASE)
+        false_regex = re.compile("^n$", re.IGNORECASE)
+
+        if not isinstance(value, str):
+            return None
+        elif true_regex.match(value):
+            return True
+        elif false_regex.match(value):
+            return False
+
+        return None
+
+    def derive_votetype(self, row):
+        """
+        Given row in voter history, make a consistent vote type.
+        """
+        true_regex = re.compile("^y$", re.IGNORECASE)
+        # false_regex = re.compile("^n$", re.IGNORECASE)
+
+        if isinstance(row["fl_absentee"], str) and true_regex.match(row["fl_absentee"]):
+            return "absentee"
```

Comment:
Later when you are inserting the data, you will want to use votetype to create absentee voting features. To do that you'll need to define another yaml field: "absentee_ballot_code: absentee".

As I was mentioning to Nick, you can get basic voting history features easily, but you do have to manually add them to the whitelists of allowed states in the feature functions. In your case, you can add west_virginia to the state lists in "get_participation", "get_local_engagement", and "get_votetype_from_dataframe" (that one is for absentee ftrs)